### PR TITLE
Goa 2122 ontology traversal rest endpoint

### PR DESCRIPTION
**Pull request by eddturner**
This request includes the REST end-point in ontology, for accessing ontology term graph traversal operations, whose underlying logic was developed in previous pull-requests (https://github.com/ebi-uniprot/QuickGOBE/pull/8 and https://github.com/ebi-uniprot/QuickGOBE/pull/9).

Specifically, there are ancestors/descendants/paths end-points, which can be accessed, and have corresponding tests.

### CODE REVIEW

**Comment by eddturner** [comment]
Review can use `AnnotationController`'s `/ancestors`, `/descendants`, `/path` endpoints as starting point.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+
+        @Test(expected = IllegalArgumentException.class)
+        public void findingPathsWithEmptyEndVerticesThrowsException() {
+            ontologyGraph.paths(
+                    ids("1"),
+                    Collections.emptySet(),
+                    CAPABLE_OF_PART_OF
+            );
+        }
+
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java]: Worth adding these?


        @Test(expected = IllegalArgumentException.class)
        public void findingPathsWithEmptyStartAndVerticesThrowsException() {
            ontologyGraph.paths(
                    Collections.emptySet(),
                    Collections.emptySet(),
                    CAPABLE_OF_PART_OF
            );
        }


        @Test(expected = IllegalArgumentException.class)
        public void findingPathsWithNullStartAndEndVerticesThrowsException() {
            ontologyGraph.paths(
                    null,
                    null,
                    CAPABLE_OF_PART_OF
            );
        }

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+            OntologyRelationship v1_CP_v2 = createRelationship(id("1"), id("2"), CAPABLE_OF_PART_OF);
+            OntologyRelationship v2_OI_v3 = createRelationship(id("2"), id("3"), OCCURS_IN);
+            ontologyGraph.addRelationships(
+                    asList(
+                            v1_CO_v2,
+                            v1_CP_v2,
+                            v2_OI_v3
+                    )
+            );
+
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java]: This setup code is duplicated quite a lot. 
Would it better to do it either once for all tests, call creation code or have one test method that does the set up and then tests all the permutations you are proving for that setup.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                    }
+                }
+            }
+
+            ancestorEdgesMap.put(vertex, ancestorEdgesOfV);
+        }
+        return ancestorEdgesMap.get(vertex);
+    }
+
+    private List<OntologyRelationship> successorEdges(
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: I think this method would benefit from some comment or javadoc even if its private.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+        fromVertices.forEach(
+                from -> {
+                    Set<OntologyRelationship> outgoingEdgesOfV = ontology.outgoingEdgesOf(from);
+                    outgoingEdgesOfV.stream()
+                            .filter(e -> relations.contains(e.relationship) && toVertices.contains(e.parent))
+                            .forEach(successors::add);
+                }
+        );
+
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: Slightly simpler

        fromVertices.forEach(
                from -> ontology.outgoingEdgesOf(from)
                        .stream()
                        .filter(e -> relations.contains(e.relationship) && toVertices.contains(e.parent))
                        .forEach(successors::add)
        );

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/controller/OBOControllerIT.java [hunk 5]
```diff
@@ -388,6 +403,219 @@ public void canRetrieveCompleteByFindAll() throws Exception {
...
+
+        ResultActions response = mockMvc.perform(
+                get(buildTermsURL(asCSV(bottom, secondBottom)) + "/" + ANCESTORS_ENDPOINT));
+
+        response.andDo(print())
+                .andExpect(jsonPath("$.numberOfHits").value(2))
+                .andExpect(jsonPath("$.results[0].ancestors", hasSize(10)))
+                .andExpect(jsonPath("$.results[1].ancestors", hasSize(9)));
+    }
+
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/controller/OBOControllerIT.java]: Is testing the number of returned results sufficient? 
Should the test validate the content to make sure it refers to the target terms?

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java (hunk 1 above)
Comment (eddturner) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java]: done

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java (hunk 4 above)
Comment (eddturner) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: done

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/controller/OBOControllerIT.java (hunk 5 above)
Comment (eddturner) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/controller/OBOControllerIT.java]: i've deferred the testing of the contents to the service layer and the tests for the graph itself; the number check is just a sanity check that the expected number of results are being returned.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java (hunk 3 above)
Comment (eddturner) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: done

**Pull request merged**